### PR TITLE
Feature/tmq

### PR DESCRIPTION
**Pull request by ryanbtp**
* [x] Use tokio / tmq
* [x] Replace protobuf with prost
* [x] Support multiple connections
* [x] Strategy for choosing write connection
* [x] tokio-console support for development builds 
* [x] Zmq publisher for embedded
* [x] Drop / close
* [x] De duplication
* [x] GC timed out reply channels
* [x] Add server socket to EmbeddedTP, test using actual ZMQ

# PR Checklist
## Please complete this checklist after opening your PR

* [x] I have updated documentation as necessary
* [x] I have updated helm chart(s) if needed
* [x] I have added tests if needed
* [x] All new and existing tests are passing
* [x] Any breaking changes are clearly flagged and documented
* [x] I’ve included a link to any relevant ticket(s)

### CODE REVIEW

File: crates/api/src/inmem.rs [hunk 1]
```diff
@@ -359,27 +377,92 @@ impl EmbeddedChronicleTp {
...
+            let task = local.run_until(async move {
+                tokio::task::spawn_local(async move {
+                    let (mut router_tx, mut router_rx) = router(&Context::new())
+                        .bind(&format!("tcp://{}", listen_addr))
+                        .unwrap()
+                        .split();
+
+                    debug!(listen_addr = ?listen_addr, "Embedded TP listening");
+                    loop {
+                        select! {
```
[the remark this answers is not in the document]
Comment (ryanbtp) [reply crates/api/src/inmem.rs:403]: This one is ok, it was the client that was using UnorderedFutures that was the
issue

File: crates/api/src/inmem.rs [hunk 2]
```diff
@@ -359,27 +383,109 @@ impl EmbeddedChronicleTp {
...
+                            message = router_rx.next().fuse() => {
+                              if message.is_none() {
+                                break;
+                              }
+
+                              let mut multipart = message.unwrap().unwrap();
+
+                              debug!(request = ?multipart);
+                              let _request = multipart.pop_front().unwrap().to_vec();
```
Comment (mtbc) [inline crates/api/src/inmem.rs:418]: It's odd that we pop two and ignore the first. Perhaps some adjective in the naming might help illustrate what's going on?

File: crates/async-sawtooth-sdk/src/error.rs [hunk 3]
```diff
@@ -1,30 +1,27 @@
...
 
-    #[error("Sawtooth receive error: {0}")]
-    ReceiveSawtooth(#[from] ReceiveError),
+    #[error("Timeout error: {0}")]
+    Elapsed(#[from] tokio::time::error::Elapsed),
 
-    #[error("Sawtooth send error: {0}")]
-    SendSawtooth(#[from] sawtooth_sdk::messaging::stream::SendError),
+    #[error("Oneshot receive error: {0}")]
+    ReceiveOneshor(#[from] tokio::sync::oneshot::error::RecvError),
```
Comment (mtbc) [inline crates/async-sawtooth-sdk/src/error.rs:19]:
```suggestion
    ReceiveOneshot(#[from] tokio::sync::oneshot::error::RecvError),
```

File: crates/async-sawtooth-sdk/src/zmq_client.rs [hunk 4]
```diff
@@ -118,168 +132,767 @@ impl RequestResponseSawtoothChannel
...
+
+                    if self.lru.get(&*message_hash).is_none() {
+                        //We still need to broadcast this message
+                        self.broadcast
+                            .send(Arc::new((message_type, message.content)))
+                            .map_err(|e| error!(failed_to_broadcast = ?e, "Failed to broadcast"))
+                            .ok();
+                    } else {
+                        trace!(ignore_duplicate=%hex::encode(message_hash));
+                        self.lru.push(message_hash.try_into().unwrap(), ());
```
Comment (mtbc) [inline crates/async-sawtooth-sdk/src/zmq_client.rs:266]:
```suggestion
                            .ok();
                        self.lru.push(message_hash.try_into().unwrap(), ());
                    } else {
                        trace!(ignore_duplicate=%hex::encode(message_hash));
```

File: crates/async-sawtooth-sdk/src/zmq_client.rs [hunk 5]
```diff
@@ -118,168 +132,767 @@ impl RequestResponseSawtoothChannel
...
+                          router.route(index,msg).await;
                         } else {
-                            break Some((response.unwrap(), channel));
+                          warn!(invalid_message = ?msg);
                         }
+                      },
+                      Some((index, Err(e))) => {
+                        error!(connection_index = index, ?e);
+                      }
+                      _ =>  {}
```
Comment (mtbc) [inline crates/async-sawtooth-sdk/src/zmq_client.rs:577]: `None` isn't at all notable?

File: docs/namespaces.md [hunk 6]
```diff
@@ -19,21 +19,21 @@ a namespace binding section in the [configuration](/config) file for this.
...
 namespace as 'default'.
 
-## Built-In Namespaces
+## Built-In namespaces
 
 ### default
 
 The default namespace is the namespace that is used when you do not specify one
-in a GraphQL mutation or query. It can be bound to any UUID, and should be set
+in a graphql mutation or query. It can be bound to any UUID, and should be set
```
Comment (mtbc) [inline docs/namespaces.md:27]: Seem to be many decapitalizations in this file, a rebase slip?

File: crates/async-sawtooth-sdk/src/zmq_client.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtbc) [reply crates/async-sawtooth-sdk/src/zmq_client.rs]: Although the diff's now outdated, it still seems odd to me that the `push` is in the `else` branch.

File: docs/namespaces.md (hunk 6 above)
Comment (ryanbtp) [reply docs/namespaces.md:27]: Yes most likely

File: crates/api/src/inmem.rs (hunk 2 above)
Comment (ryanbtp) [reply crates/api/src/inmem.rs]: Blank message frames are meant to be used in mulipart by routers according to zmq

Comment (mtbc) [reply crates/api/src/inmem.rs]: Ah, maybe call it `_blank_request` or somesuch?

File: .VERSION [hunk 7]
```diff
@@ -0,0 +1 @@
+v0.7.4-27-gc8a21fc0
```
Comment (mtbc) [inline .VERSION:1]: Do we want to be committing this?

Comment (ryanbtp) [reply .VERSION]: No, and it's in .gitignore. Removed

File: crates/common/src/opa.rs [hunk 8]
```diff
@@ -533,6 +541,7 @@ mod tests {
     }
 
     #[tokio::test]
+    #[ignore]
```
Comment (mtbc) [inline crates/common/src/opa.rs:544]: This is worrying. Should we add a `reason` or open a ticket or somesuch?

Comment (ryanbtp) [reply crates/common/src/opa.rs:544]: This was due to rebasing onto broken .rego, now unignored

Also, in future when asserting a Result<>, just use unwrap() as it prints the error

**Review by mtbc — APPROVED** [review]

File: Cargo.toml [hunk 9]
```diff
@@ -92,9 +95,12 @@ owo-colors = "3.5.0"
 parking_lot = "0.12.0"
 percent-encoding = "2.1.0"
 pin-project = "1.0.12"
+pin-project-lite = "*"
```
Comment (scealiontach) [inline Cargo.toml:98]: We should not be using "*" dependencies

**Review by scealiontach — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 9 above)
Comment (ryanbtp) [reply Cargo.toml]: Fixed

**Review by suchapalaver — APPROVED** [review]
It's a ton of work! I feel like you could use commit messages to parse this up so it's easier for people to break down the changes - it may as well be one commit, really. But that's more for my benefit ;)

File: Cargo.toml (hunk 9 above)
Comment (mtbc) [reply Cargo.toml]: There still seem to be some in the pushed version.

**Review by scealiontach — APPROVED** [review]

**Pull request merged**